AppController/djinn.rb

```diff
@@ -4927,6 +4855,40 @@ def find_lowest_free_port(starting_port)
   end
 
 
+  # Updates @app_info_map with registered instances.
+  def update_registered_instances(version_key)
+    begin
+      zk_instances = ZKInterface.get_children(
+        "/appscale/instances_by_version/#{version_key}")
+    rescue FailedZooKeeperOperationException
+      Djinn.log_error('Unable to update registered instances.')
+      return
+    end
+
+    unless zk_instances.empty?
+      @app_info_map[version_key] = {} unless @app_info_map.key?(version_key)
+      unless @app_info_map[version_key].key?('appservers')
+        @app_info_map[version_key]['appservers'] = []
+      end
+    end
+    known_instances = @app_info_map[version_key]['appservers']
+
+    zk_instances.each { |instance_key|
+      next if known_instances.include?(instance_key)
+
+      # Find and remove an entry for this AppServer node and app.
+      ip = instance_key.split(':')[0]
+      match = known_instances.index("#{ip}:-1")
+      if match
+        known_instances.delete_at(match)
+      else
+        Djinn.log_warn("Adding an unassigned instance: #{instance_key}.")
+      end
+      known_instances << instance_key
+    }
+  end
```

Comment:
I think here we should also do the _other side of the coin_ check: anything in app_info_map but not registered, shouldn't it be removed? There is something like that for the compute node in check_running_appservers, also that part I think is redundant since now the AMS is doing the check of what is running locally (ie less interaction with Monit from different processes).

Reply:
The AppManager is not yet checking what is running locally (at least not on a regular basis). The router is still responsible for detecting failed instances and mutating app_info_map state.

I was considering this block to just be a direct replacement for the add_routing call to minimize changes. I was hoping to avoid a huge diff. However, I see now that instances could be erroneously re-added to app_info_map if the AppManager does not remove the instance registration entry before the next duty cycle. I'll probably need a separate pull request before this one to separate assignment and registration structures.

Reply:
I see. But couldn't we give the shadow the biz of checking what's running? I guess my question is: do you believe to be monit or haproxy the best check for the failure or well-being of AppServers?

Reply:
I think the AppManager should be the one checking instance health. I don't think the autoscaler should be able to start and stop individual instances. It should just assign a count per version to each machine. Each compute machine should be responsible for the details about which individual instances should be running.

The AppManager could still use HAProxy as one condition of its health check.